design-docs/007-basic-expiration.md

```diff
@@ -0,0 +1,561 @@
+
+Icechunk keeps exhaustive track of history. Currently every snapshot ever written is available
+in object store and never deleted. Many times, that is not what users want. Our experience with
+Arraylake has taught us that 
+users usually want to keep "some history" but not all of it. The reason is:
+wasted storage. If they never go back to check previous versions, why would they want to pay
+for the extra storage.
+
+There are many ways to select what versions (snapshots) should be kept
+around and which ones should be deleted. In our experience, a good balance of
+simplicity and power is: _Expire snapshots older than certain threshold_.
+
+The expiration process has two stages: history edition (aka expiration)
+and garbage collection.
+
+During history edition, references and snapshots can be edited and added
+in a way that generates a different view of the history of the repository.
+During this process no objects are deleted. Some may be edited, but in a
+non destructive way.
+
+Garbage collection is the process by which objects that are no longer needed
+get deleted, freeing storage space. In our case, we declare that the object
+that are no longer needed are those that cannot be reached from any tags or branches.
+In this sense, tags and branches are what keep snapshots alive by pointing at them.
+
+In this way, if history edition generates unreachable snapshots, then garbage
+collection can free them. Importantly, once a snapshot becomes unreachable,
+its exclusive manifests and its exclusive chunks are also unreachable. This
+is where the storage savings actually happen, by deleting chunks that were
+overwritten (or flagged as deleted) in later versions of the repository.
+
+Keeping history is still very useful: it provides the infrastructure for
+transactions, it allows users to undo mistakes, improves auditability, and,
+in certain cases, it's a core feature users need (example: insurance industry).
+
+
+## The simpler problem of snapshot removal
+
+To understand how to expire old snapshots, we first need to understand
+how we can abandon a series of arbitrary snapshots.
+
+We present a diagram of a repository snapshots. Arrows follow the `parent`
+relationship. Above in the diagram is the original repo, and below, we show
+two possible ways to abandon snapshots `X1, X2, ..., Xn`
+
+```mermaid
+stateDiagram-v2
+  direction TB
+  state Original {
+
+    state "X1, X2, ..., Xn" as X
+    state "..." as REST
+
+    [*] --> N1
+    N1 --> N2
+    N2 --> A
+    A --> X
+    X --> B
+    B --> REST
+    
+  }
+
+  Original --> Skip
+  Original --> EditParents
+
+  state "Update A in place" as Skip {
+
+    state "N1" as N1E
+    state "N2" as N2E
+    state "A (updated)" as AE
+    state "B" as BE
+    state "X1, X2, ..., Xn" as XE
+    state "..." as RESTE   %% codespell:ignore
+
+    [*] --> N1E
+    N1E --> N2E
+    N2E --> AE
+    AE --> BE
+    XE --> BE
+    BE --> RESTE    %% codespell:ignore
```

Comment:
What is `RESTE` and `RESTP` (below)

Reply:
I need to give the nodes a different name for the graph to work ... basically a whole other series of commits. Will add the explanation.

Reply:
Ahh I see, the label is wrong. Thank you.

Reply:
done